Add new feature to Apple TV platform

postlund opened the pull request:
## Description:
This PR updates pyatv to 0.3.2 and adds new features, mainly:

- Support for "stop"
- Remote control buttons (e.g. menu, arrow keys)
- Device authentication needed for play_media on tvOS 10.2+
- Scanning for devices

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2858

## Example entry for `configuration.yaml` (if applicable):
```yaml
apple_tv:
  - name: Apple TV
    host: 10.0.10.20
    login_id: 00000000-1234-5678-9012-345678901234
    start_off: true
    credentials: 8660DEA5154FB46B:20B94847926112B3F46F85DB3A7311830463BF65570C22C3786E27F38C3326CF
```
``credentials`` is new and user by the device authentication feature.

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

postlund committed:
Lift Apple TV to pyatv 0.3.2

Update code to use basic new features.

postlund committed:
Support button presses in Apple TV

postlund committed:
Support device authentication

pvizeli commented:
It is only a question. Could be make sense to add this as remote platform?

postlund commented:
Been thinking a bit about what you said and it's starting to make sense. It would probably be a good idea. I assume you mean extracting apple_tv to its own component and then load a remote and a media_player platform when a new device is added, no?

balloob commented:
@postlund yep.

postlund committed:
Convert Apple TV to a component

A media_player platform and a remote platform will be loaded for each
manually configured or discovered device.

postlund commented:
@pvizeli @balloob I have finished the code changes by converting it to a component with two platforms now now. Just have to update documentation which will require some rearrangement.

balloob commented:
Can you update `.coveragerc` and re-run `script/gen_requirements_all.py` ?

balloob commented:
Also, please include in the description what the breaking change entails so I can copy paste that to the release notes.

balloob commented on homeassistant/components/media_player/apple_tv.py:
```diff
@@ -6,134 +6,179 @@
-    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-    vol.Optional(CONF_START_OFF, default=False): cv.boolean
-})
+def request_configuration(hass, config, atv, credentials):
```
Is authentication something that we should move to the component too?

postlund replied:
It probably can, but it depends on how we want to handle entity lookups in the service handler. Since entity ids passed are passed to apple_tv_authenticate (like with all services), the entity objects must be saved in the media_player platform (which is done now) but retrieved from the component. Not an implementation problem in any way, more a technical thing if it's "OK" to have the component-platform-coupling this way?

balloob replied:
Yes, coupling between platforms and components that share auth is fine as long as they share the same name 👍

postlund committed:
Move device auth to apple_tv component

postlund committed:
Update requirements and coverage config

postlund commented:
I have moved the device authentication service to the component now and fixed the other remarks. Hopefully everything will pass now. As soon as I get a few minutes over I will try to fix the documentation and write description about the breaking change. Feel free to leave comments on the code until I have time to finish the rest!

postlund committed:
Add scan support to apple_tv

postlund commented:
I added an additional service for scanning for devices. It makes it easier for user to find them and the login_id as they don't have to use atvremote from CLI.

postlund commented:
Everything should be in place now (including documentation). A suitable text for breaking changes might be along the lines:

Previously, Apple TV supported was implemented as a ``media_player``. It has now been converted into a component and must be configured as such:
```yaml
apple_tv:
  - host: 10.0.10.201
    name: Apple TV 1
    login_id: 00000000-1234-5678-9012-345678901234)
  - host: 10.0.10.202
    name: Apple TV 2
    login_id: 00000000-1234-5678-9012-345678901432)
```

pvizeli reviewed (approved):
Looks good to me. @balloob can you also verify that?
